File: simple-stack-example-mvp-view/src/main/java/com/zhuinden/simplestackdemoexamplemvp/presentation/paths/addoredittask/AddOrEditTaskPresenter.java
package com.zhuinden.simplestackdemoexamplemvp.presentation.paths.addoredittask;

import android.support.annotation.Nullable;

import com.zhuinden.simplestack.Backstack;
import com.zhuinden.simplestack.Bundleable;
import com.zhuinden.simplestack.History;
import com.zhuinden.simplestack.StateChange;
import com.zhuinden.simplestackdemoexamplemvp.data.repository.TaskRepository;
import com.zhuinden.simplestackdemoexamplemvp.presentation.objects.Task;
import com.zhuinden.simplestackdemoexamplemvp.presentation.paths.tasks.TasksKey;
import com.zhuinden.simplestackdemoexamplemvp.presentation.paths.tasks.TasksView;
import com.zhuinden.simplestackdemoexamplemvp.util.BasePresenter;
import com.zhuinden.simplestackdemoexamplemvp.util.MessageQueue;
import com.zhuinden.simplestackdemoexamplemvp.util.Strings;
import com.zhuinden.statebundle.StateBundle;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;


/**
 * Created by Owner on 2017. 01. 27..
 */
// UNSCOPED
public class AddOrEditTaskPresenter
        extends BasePresenter<AddOrEditTaskView>
        implements Bundleable {
    String title;
    String description;

    private final TaskRepository taskRepository;
    private final MessageQueue messageQueue;
    private final Backstack backstack;

    @Inject
    public AddOrEditTaskPresenter(TaskRepository taskRepository, MessageQueue messageQueue, Backstack backstack) {
        this.taskRepository = taskRepository;
        this.messageQueue = messageQueue;
        this.backstack = backstack;
    }

    public void updateTitle(String title) {
        this.title = title;
    }

    public void updateDescription(String description) {
        this.description = description;
    }

    String taskId;

    Task task;

    @Override
    protected void onAttach(AddOrEditTaskView view) {
        AddOrEditTaskKey addOrEditTaskKey = Backstack.getKey(view.getContext());
        taskId = addOrEditTaskKey.taskId();
        if(!Strings.isNullOrEmpty(taskId)) {
            taskRepository.findTask(addOrEditTaskKey.taskId()).observeOn(AndroidSchedulers.mainThread()).subscribe(taskOptional -> {
                if(taskOptional.isPresent()) {
                    task = taskOptional.get();
                    if(this.title == null || this.description == null) {
                        this.title = task.title();
                        this.description = task.description();
                        view.setTitle(title);
                        view.setDescription(description);
                    }
                }
            });
        }
    }

    @Override
    protected void onDetach(AddOrEditTaskView view) {

    }

    public StateBundle toBundle() {
        StateBundle bundle = new StateBundle();
        bundle.putString("title", title);
        bundle.putString("description", description);
        return bundle;
    }

    @Override
    public void fromBundle(@Nullable StateBundle bundle) {
        if(bundle != null) {
            title = bundle.getString("title");
            description = bundle.getString("description");
        }
    }

    public void saveTask() {
        if(!Strings.isNullOrEmpty(title) && !Strings.isNullOrEmpty(description)) {
            taskRepository.insertTask(task == null ? Task.createNewActiveTask(title, description) : task.toBuilder()
                    .setTitle(title)
                    .setDescription(description)
                    .build());
        }
    }

    public void navigateBack() {
        AddOrEditTaskKey addOrEditTaskKey = Backstack.getKey(getView().getContext());
        if(addOrEditTaskKey.parent() instanceof TasksKey) {
            messageQueue.pushMessageTo(addOrEditTaskKey.parent(), new TasksView.SavedSuccessfullyMessage());
            backstack.goBack();
        } else {
            backstack.setHistory(History.builderFrom(backstack).removeUntil(TasksKey.create()).build(), StateChange.BACKWARD);
        }
    }
}

File: simple-stack-example-mvp-view/src/main/java/com/zhuinden/simplestackdemoexamplemvp/presentation/paths/addoredittask/AddOrEditTaskPresenter.kt
package com.zhuinden.simplestackdemoexamplemvp.presentation.paths.addoredittask

import com.zhuinden.simplestack.Backstack
import com.zhuinden.simplestack.Bundleable
import com.zhuinden.simplestack.StateChange
import com.zhuinden.simplestackdemoexamplemvp.data.repository.TaskRepository
import com.zhuinden.simplestackdemoexamplemvp.presentation.objects.Task
import com.zhuinden.simplestackdemoexamplemvp.presentation.paths.tasks.TasksKey
import com.zhuinden.simplestackdemoexamplemvp.presentation.paths.tasks.TasksView
import com.zhuinden.simplestackdemoexamplemvp.util.BasePresenter
import com.zhuinden.simplestackdemoexamplemvp.util.MessageQueue
import com.zhuinden.simplestackdemoexamplemvp.util.hideKeyboard
import com.zhuinden.statebundle.StateBundle
import io.reactivex.android.schedulers.AndroidSchedulers
import javax.inject.Inject


/**
 * Created by Owner on 2017. 01. 27..
 */
// UNSCOPED
class AddOrEditTaskPresenter @Inject constructor(
    private val taskRepository: TaskRepository,
    private val messageQueue: MessageQueue,
    private val backstack: Backstack
) : BasePresenter<AddOrEditTaskView>(), Bundleable {
    private var title: String? = null
    private var description: String? = null

    private var taskId: String = ""

    private var task: Task? = null

    fun updateTitle(title: String) {
        this.title = title
    }

    fun updateDescription(description: String) {
        this.description = description
    }

    override fun onAttach(view: AddOrEditTaskView) {
        val addOrEditTaskKey = Backstack.getKey<AddOrEditTaskKey>(view.context)
        taskId = addOrEditTaskKey.taskId

        if (taskId.isNotEmpty()) {
            taskRepository.findTask(addOrEditTaskKey.taskId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { taskOptional ->
                    val task = taskOptional.takeIf { it.isPresent }?.get() ?: return@subscribe
                    this.task = task

                    if (this.title == null || this.description == null) {
                        this.title = task.title
                        this.description = task.description
                        view.setTitle(title ?: "")
                        view.setDescription(description ?: "")
                    }
                }
        }
    }

    override fun onDetach(view: AddOrEditTaskView) {

    }

    override fun toBundle(): StateBundle = StateBundle().apply {
        putString("title", title)
        putString("description", description)
    }

    override fun fromBundle(bundle: StateBundle?) {
        bundle?.run {
            title = getString("title")
            description = getString("description")
        }
    }

    fun saveTask(): Boolean {
        val title = title
        val description = description
        val task = task

        if (title.isNullOrEmpty() || description.isNullOrEmpty()) {
            return false
        }

        taskRepository.insertTask(when {
            taskId.isEmpty() -> {
                Task.createNewActiveTask(title!!, description!!)
            }
            else -> {
                task?.copy(title = title, description = description) ?: return false
            }
        })
        return true

    }

    fun navigateBack() {
        val addOrEditTaskKey = Backstack.getKey<AddOrEditTaskKey>(view!!.context)
        view!!.hideKeyboard()
        when(addOrEditTaskKey) {
            is AddOrEditTaskKey.AddTaskKey -> {
                messageQueue.pushMessageTo(backstack.root<TasksKey>()!!, TasksView.SavedSuccessfullyMessage())
                backstack.goBack()
            }
            is AddOrEditTaskKey.EditTaskKey -> {
                backstack.jumpToRoot(StateChange.BACKWARD)
            }
        }
    }
}
